lib/usd/ui/views/TreeModel.cpp

```diff
@@ -0,0 +1,239 @@
+//
+// Copyright 2019 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "TreeModel.h"
+#include "TreeItem.h"
+#include "ItemDelegate.h"
+
+#include <QtWidgets/QTreeView>
+#include <QtCore/QSortFilterProxyModel>
+
+MAYAUSD_NS_DEF {
+
+TreeModel::TreeModel(QObject* parent) noexcept
+	: ParentClass{ parent }
+{
+}
+
+QVariant TreeModel::headerData(int section, Qt::Orientation orientation, int role /*= Qt::DisplayRole*/) const
+{
+	if (orientation == Qt::Horizontal)
+	{
+		switch (role)
+		{
+		case Qt::CheckStateRole:
+			if (section == kTreeColumn_Load)
+			{
+				return QVariant::fromValue<int>(Qt::Checked);
+			}
+			break;
+		}
+	}
+	return ParentClass::headerData(section, orientation, role);
+}
+
+QVariant TreeModel::data(const QModelIndex &index, int role /*= Qt::DisplayRole*/) const
+{
+	if (!index.isValid())
+		return QVariant();
+
+	TreeItem* item = static_cast<TreeItem*>(itemFromIndex(index));
+
+	if ((role == Qt::CheckStateRole) && (index.column() == kTreeColumn_Load))
+		return item->checkState();
+
+	return ParentClass::data(index, role);
+}
+
+Qt::ItemFlags TreeModel::flags(const QModelIndex &index) const
+{
+	if (!index.isValid())
+		return 0;
+
+	Qt::ItemFlags flags = Qt::ItemIsEnabled | Qt::ItemIsSelectable;
+	if (index.column() == kTreeColumn_Load)
+		flags |= Qt::ItemIsUserCheckable;
+	else
+		flags = ParentClass::flags(index);
+
+	return flags;
+}
+
+bool TreeModel::setData(const QModelIndex &index, const QVariant &value, int role /*= Qt::EditRole*/)
+{
+	if (index.column() == kTreeColumn_Load)
+	{
+		if (role == Qt::EditRole)
+			return false;
+
+		if (role == Qt::CheckStateRole)
+		{
+			TreeItem* item = static_cast<TreeItem*>(itemFromIndex(index));
+			item->setCheckState(static_cast<Qt::CheckState>(value.toInt()));
+			emit dataChanged(index, index);
+
+			// If we 'unchecked' this item, we need to uncheck all its children.
+			// If we 'checked' this item, we need to check all of the parent items
+			// and all of the child items.
+			if (item->checkState() == Qt::Unchecked) {
+				setChildCheckState(index, Qt::Unchecked);
+			} else {
+				setParentsCheckState(index, Qt::Checked);
+				setChildCheckState(index, Qt::Checked);
+			}
+			return true;
+		}
+	}
+
+	return ParentClass::setData(index, value, role);
+}
+
+void TreeModel::setParentsCheckState(const QModelIndex &child, Qt::CheckState state)
```

Comment:
This method (and next one) are used to check/uncheck items parent/child items when the user interactively checks/unchecks an item in the tree.